packages/snaps-utils/src/url.ts

```diff
@@ -0,0 +1,93 @@
+import { type SnapId } from '@metamask/snaps-sdk';
+import { type Infer, is, pattern, string } from '@metamask/superstruct';
+import { assert } from '@metamask/utils';
+
+import { assertIsValidSnapId, stripSnapPrefix } from './snaps';
+
+export const METAMASK_URL_REGEX =
+  /^metamask:\/\/(?<authority>(client|snap))(?<path>.+)$/u;
+
+export const MetaMaskUrlStruct = pattern(string(), METAMASK_URL_REGEX);
+
+export type MetaMaskUrl = Infer<typeof MetaMaskUrlStruct>;
+
+export type Authority = 'client' | 'snap';
+
+export const ExtensionPaths = ['/'];
+
+export const SnapPaths = ['/home'];
+
+/**
+ * Parse a url string to an object containing the authority and path.
+ * This validates the url before parsing it.
+ *
+ * @param url - The url to validate and parse.
+ * @returns The parsed url.
+ * @throws If the authority or path is invalid.
+ */
+export function parseMetaMaskUrl(url: MetaMaskUrl): {
```

Comment:
If you use URL escape sequences you could simplify the parsing quite a bit using `new URL(string)`

Reply:
Not sure I follow?

Reply:
You could use `url.hostname` and `url.pathname` instead of parsing regex which might be more prone to unforeseen edge-cases than battle-tested `URL`

Reply:
Ok I agree with the usage of `URL`, but I don't see how the escape sequences make the logic any simpler. The url still works with the unescaped sequences and in fact the logic may be even more complex if I'm having to convert the escaped sequences back to normal.

Reply:
made changes here: [c672e2b](https://github.com/MetaMask/snaps/pull/2719/commits/c672e2b3421b753165ac297f5719b1b1f38caa09)